modules/ingester/instance.go

```diff
@@ -376,27 +376,33 @@ func (i *instance) GetBlockToBeFlushed(blockID uuid.UUID) *LocalBlock {
 	return nil
 }
 
-func (i *instance) ClearFlushedBlocks(completeBlockTimeout time.Duration) error {
+func (i *instance) ClearOldBlocks(flushObjectStorage bool, completeBlockTimeout time.Duration) error {
 	var err error
 
 	i.blocksMtx.Lock()
 	defer i.blocksMtx.Unlock()
 
+	cutoff := time.Now().Add(-completeBlockTimeout)
+
 	for idx, b := range i.completeBlocks {
-		flushedTime := b.FlushedTime()
-		if flushedTime.IsZero() {
+		// If we are flushing to object storage, never delete blocks that have not been flushed.
+		if flushObjectStorage && b.FlushedTime().IsZero() {
 			continue
 		}
 
-		if flushedTime.Add(completeBlockTimeout).Before(time.Now()) {
-			i.completeBlocks = append(i.completeBlocks[:idx], i.completeBlocks[idx+1:]...)
+		// If the block contains data within retention, keep it.
+		if b.BlockMeta().EndTime.After(cutoff) {
+			continue
+		}
 
-			err = i.local.ClearBlock((uuid.UUID)(b.BlockMeta().BlockID), i.instanceID)
-			if err == nil {
-				metricBlocksClearedTotal.Inc()
-			}
-			break
+		// This block can be deleted.
+		i.completeBlocks = append(i.completeBlocks[:idx], i.completeBlocks[idx+1:]...)
```

Comment:
Should this line be after the clearBlock try? If we remove the id from the list of complete blocks and then we get an error, the block will be dangling forever

Reply:
Good catch. I guess that has been an issue for a long time. Also, the previous code was only deleting 1 block per pass, and I attemped to delete all matching blocks. This was broken and needed some other updates because we are modifying the slice while iterating it. Thoughts?  There are several ways to do this, and none are particularly clear.